### PR TITLE
Failure injector plugin implementation.

### PR DESCRIPTION
**Pull request by mashhurs**
## What does this PR do?
Introduces the `failure_injector` integration plugin to test the Logstash health. I don't see the input plugin type need, so filter and output plugin types are considered.
Usage: plugin has `degrade_at` (array) and `crash_at` (string) configurations define the phases `failure_injector` needs to simulate degrade or crash behavior.

Example config:
```
filter {
    failure_injector {
        #degrade_at => [register, filter, close] # receives one or any of [register, filter, close] 
        crash_at => close # [register, filter, close]
    }
}
output {
    failure_injector {
        degrade_at => [receive] # accepts one or any of [register, receive, close]
        #crash_at => receive # accepts one of [register, receive, close]
    }
}
```

## Why is it important/What is the impact to the user?
N.A

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [x] Check the `unknown` status when plugin crashes at register or close phase (filter or receive phase causes `TERMINATED` which is our expectation)

## How to test this PR locally
- This PR needs to be tested with https://github.com/elastic/logstash/pull/16398
- Install the plugin with `bin/logstash-plugin install qa/support/logstash-integration-failure_injector/logstash-integration-failure_injector-0.0.1.gem` command
- Use following config and hit the `GET /_health_report` endpoint
```
# pipelines.yml
- pipeline.id: slow-start-pp
  config.string: |
    input { heartbeat {} }
    filter { failure_injector { degrade_at => [register] } }
    output { stdout {} }
- pipeline.id: filter-crashed-pp
  config.string: |
    input { generator { count => 10 } }
    filter { failure_injector { crash_at => register } }
    output { stdout {} }
- pipeline.id: normal-heavy-finish-pp
  config.string: |
    input { generator { count => 10 } }
    filter { failure_injector { degrade_at => filter } }
    output { failure_injector { degrade_at => receive } }
- pipeline.id: output-crashed-pp
  config.string: |
    input { generator { count => 10 } }
    output { failure_injector { crash_at => receive } }
- pipeline.id: pp-to-pp-upstream
  config.string: input { generator {} } output { pipeline { send_to => [myVirtualAddress] } }
- pipeline.id: pp-to-pp-downstream
  config.string: input { pipeline { address => myVirtualAddress } } output { failure_injector { crash_at => receive } }
```
- See the log section for the result

## Logs

```
# Output
{
    "status": "red",
    "symptom": "1 indicator is unhealthy (`pipelines`)",
    "indicators": {
        "pipelines": {
            "status": "red",
            "symptom": "2 indicators are unhealthy (`pp-to-pp-downstream`, `output-crashed-pp`), 1 indicator is degraded or at risk of becoming unhealthy (`normal-heavy-finish-pp`), 2 indicators are healthy (`slow-start-pp`, `pp-to-pp-upstream`), and 1 indicator is unknown (`filter-crashed-pp`)",
            "indicators": {
                "filter-crashed-pp": {
                    "status": "unknown",
                    "symptom": "The pipeline is unknown; 1 area is impacted and 1 diagnosis is available.",
                    "diagnosis": [
                        {
                            "cause": "pipeline is not known; it may have been recently deleted",
                            "action": "check logs",
                            "help_url": null
                        }
                    ],
                    "impacts": [
                        {
                            "severity": 0,
                            "description": "pipeline is not known",
                            "impactAreas": [
                                "pipeline_execution"
                            ]
                        }
                    ],
                    "details": {
                        "run_state": "UNKNOWN"
                    }
                },
                "pp-to-pp-downstream": {
                    "status": "red",
                    "symptom": "The pipeline is unhealthy; 1 area is impacted and 1 diagnosis is available.",
                    "diagnosis": [
                        {
                            "cause": "pipeline is not running, likely because it has encountered an error",
                            "action": "check logs",
                            "help_url": null
                        }
                    ],
                    "impacts": [
                        {
                            "severity": 0,
                            "description": "pipeline is not running",
                            "impactAreas": [
                                "pipeline_execution"
                            ]
                        }
                    ],
                    "details": {
                        "run_state": "TERMINATED"
                    }
                },
                "slow-start-pp": {
                    "status": "green",
                    "symptom": "The pipeline is healthy; 0 areas are impacted and 0 diagnoses are available.",
                    "details": {
                        "run_state": "RUNNING"
                    }
                },
                "output-crashed-pp": {
                    "status": "red",
                    "symptom": "The pipeline is unhealthy; 1 area is impacted and 1 diagnosis is available.",
                    "diagnosis": [
                        {
                            "cause": "pipeline is not running, likely because it has encountered an error",
                            "action": "check logs",
                            "help_url": null
                        }
                    ],
                    "impacts": [
                        {
                            "severity": 0,
                            "description": "pipeline is not running",
                            "impactAreas": [
                                "pipeline_execution"
                            ]
                        }
                    ],
                    "details": {
                        "run_state": "TERMINATED"
                    }
                },
                "normal-heavy-finish-pp": {
                    "status": "yellow",
                    "symptom": "The pipeline is degraded or at risk of becoming unhealthy; 1 area is impacted and 1 diagnosis is available.",
                    "diagnosis": [
                        {
                            "cause": "pipeline is loading",
                            "action": "check logs",
                            "help_url": null
                        }
                    ],
                    "impacts": [
                        {
                            "severity": 1,
                            "description": "pipeline is loading",
                            "impactAreas": [
                                "pipeline_execution"
                            ]
                        }
                    ],
                    "details": {
                        "run_state": "LOADING"
                    }
                },
                "pp-to-pp-upstream": {
                    "status": "green",
                    "symptom": "The pipeline is healthy; 0 areas are impacted and 0 diagnoses are available.",
                    "details": {
                        "run_state": "RUNNING"
                    }
                }
            }
        }
    }
}
```

### CODE REVIEW

File: qa/support/logstash-integration-failure_injector/lib/logstash/outputs/failure_injector.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    @logger.debug("Plugin is closing")
+    degrade_or_crash_if_required('close')
+  end
+
+  def degrade_or_crash_if_required(phase)
+    degrate(phase) if @degrade_at.include?(phase)
+    crash(phase) if @crash_at && @crash_at == phase
+  end
+
+  def degrate(phase)
```
Comment (andsel) [inline qa/support/logstash-integration-failure_injector/lib/logstash/outputs/failure_injector.rb:60]:
```suggestion
  def degrade(phase)
```

File: qa/support/logstash-integration-failure_injector/lib/logstash/filters/failure_injector.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    @logger.debug("Plugin is closing")
+    degrade_or_crash_if_required('close')
+  end
+
+  def degrade_or_crash_if_required(phase)
+    degrate(phase) if @degrade_at.include?(phase)
+    crash(phase) if @crash_at && @crash_at == phase
+  end
+
+  def degrate(phase)
```
Comment (andsel) [inline qa/support/logstash-integration-failure_injector/lib/logstash/filters/failure_injector.rb:60]:
```suggestion
  def degrade(phase)
```

File: qa/support/logstash-integration-failure_injector/lib/logstash/outputs/failure_injector.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+require 'logstash/inputs/base'
+require 'logstash/namespace'
+
+require_relative '../utils/failure_injector_util'
+
+class LogStash::Outputs::FailureInjector < LogStash::Outputs::Base
+
+  config_name "failure_injector"
+
+  # one or any of [register, receive, close]
```
Comment (andsel) [inline qa/support/logstash-integration-failure_injector/lib/logstash/outputs/failure_injector.rb:29]: The intent of `degrate_at` and also for `crash_at` should state that these are the various phases of a plugin lifecycle

File: qa/support/logstash-integration-failure_injector/spec/unit/failure_injector_spec.rb [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      describe '#receive' do
+        it_behaves_like 'a phase that can degrade or crash', 'receive'
+      end
+    end
+
+    describe '#close' do
+      it_behaves_like 'a phase that can degrade or crash', 'close'
+    end
+
+    describe '#degrate' do
```
Comment (andsel) [inline qa/support/logstash-integration-failure_injector/spec/unit/failure_injector_spec.rb:115]:
```suggestion
    describe '#degrade' do
```

**Review by andsel — CHANGES_REQUESTED** [review]
Left some comments.
I think `decrate` is a typo, should be `degrade`.

File: qa/support/logstash-integration-failure_injector/lib/logstash/outputs/failure_injector.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    degrade_or_crash_if_required('receive')
+  end
+
+  def close
+    @logger.debug("Plugin is closing")
+    degrade_or_crash_if_required('close')
+  end
+
+  def degrade_or_crash_if_required(phase)
+    degrate(phase) if @degrade_at.include?(phase)
```
Comment (mashhurs) [inline qa/support/logstash-integration-failure_injector/lib/logstash/outputs/failure_injector.rb:56]:
```suggestion
    degrade(phase) if @degrade_at.include?(phase)
```

File: qa/support/logstash-integration-failure_injector/spec/unit/failure_injector_spec.rb [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      it { is_expected.to be <= (plugin_type == 'filter' ? LogStash::Filters::Base : LogStash::Outputs::Base) }
+      it { is_expected.to have_attributes(:config_name => "failure_injector") }
+    end
+
+    shared_examples 'a phase that can degrade or crash' do |phase|
+      context "when degrades at #{phase}" do
+        let(:params) { { 'degrade_at' => [phase] } }
+
+        it 'calls the degrate method' do
+          expect(plugin).to receive(:degrate).with(phase)
```
Comment (mashhurs) [inline qa/support/logstash-integration-failure_injector/spec/unit/failure_injector_spec.rb:48]:
```suggestion
        it 'calls the degrade method' do
          expect(plugin).to receive(:degrade).with(phase)
```

File: qa/support/logstash-integration-failure_injector/spec/unit/failure_injector_spec.rb [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    end
+
+    describe '#close' do
+      it_behaves_like 'a phase that can degrade or crash', 'close'
+    end
+
+    describe '#degrate' do
+      it 'sleeps for a certain period of time' do
+        expect(plugin).to receive(:sleep).at_least(:once)
+        plugin.degrate('filter')
```
Comment (mashhurs) [inline qa/support/logstash-integration-failure_injector/spec/unit/failure_injector_spec.rb:118]:
```suggestion
        plugin.degrade('filter')
```

File: qa/support/logstash-integration-failure_injector/lib/logstash/filters/failure_injector.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    degrade_or_crash_if_required('filter')
+  end
+
+  def close
+    @logger.debug("Plugin is closing")
+    degrade_or_crash_if_required('close')
+  end
+
+  def degrade_or_crash_if_required(phase)
+    degrate(phase) if @degrade_at.include?(phase)
```
Comment (mashhurs) [inline qa/support/logstash-integration-failure_injector/lib/logstash/filters/failure_injector.rb:56]:
```suggestion
    degrade(phase) if @degrade_at.include?(phase)
```

**Review by andsel — CHANGES_REQUESTED** [review]
Still missing a reply on https://github.com/elastic/logstash/pull/16466/files#r1783153990

File: qa/support/logstash-integration-failure_injector/lib/logstash/outputs/failure_injector.rb (hunk 3 above)
Comment (andsel) [reply qa/support/logstash-integration-failure_injector/lib/logstash/outputs/failure_injector.rb:29]: Maybe a description that cite that correspond to the phases of a plugin would be helpful, I think.

Comment (mashhurs) [reply qa/support/logstash-integration-failure_injector/lib/logstash/outputs/failure_injector.rb]: Yup, sorry missed it and addressed with [this commit](https://github.com/elastic/logstash/pull/16466/commits/4dfd471817ac2ea1d270c159b4ec6b6f5e0674dd).

**Review by andsel — APPROVED** [review]
LGTM

**Pull request merged**